### DISCUSSION TITLE
Rendering pkl durations into yaml

### DISCUSSION THREAD

**Discussion by tuck1s** [apple/pkl#371] [Q&A] [answered]
The [Duration](https://pkl-lang.org/main/current/language-reference/index.html#durations) type is great for validation. I'd like to render it to yaml, I assume an output converter is needed?
The preferred format would be as a string in seconds or `?d?h?m?s` formatted time, eg. 1h30m. Is that feasible?

**Comment by bioball** [comment, marked as answer]
Yup, an output converter is exactly where to define this!

This seems like what you want:

```
output {
  renderer = new YamlRenderer {
    converters {
      [Duration] = (it) ->
        let (days = it.toUnit("d").value.toInt())
        let (hours = it.toUnit("h").value.toInt() % 24)
        let (minutes = it.toUnit("min").value.toInt() % 60)
        let (seconds = it.toUnit("s").value % 60)
          new Listing {
            when (days != 0) {
              days
              "d"
            }
            when (hours != 0) {
              hours
              "h"
            }
            when (minutes != 0) {
              minutes
              "m"
            }
            when (seconds != 0) {
              seconds
              "s"
            }
          }.join("")
    }
  }
}

```

**Reply by tuck1s** [reply]
Thank you, that works well. I have multiple yaml output files, so I have several `output { }` sections already. Is it possible to create a renderer once, and use it for each one?

**Reply by bioball** [reply]
You can have them all amend a base module, and have the base module define the renderer.

For example, take a look at what we do in k8s:

https://github.com/apple/pkl-k8s/blob/main/generated-package/K8sObject.pkl#L146

**Reply by tuck1s** [reply]
That link is inaccessible to me, could you share a short example here?

**Reply by bioball** [reply]
Sorry, link updated!
